enhancements/network/ingress-logging-api.md

```diff
@@ -0,0 +1,364 @@
+	// request.
+	//
+	// If this field is empty, log messages use the implementation's default
+	// HTTP log format.  For HAProxy's default HTTP log format, see the
+	// HAProxy documentation:
+	// http://cbonte.github.io/haproxy-dconv/1.7/configuration.html#8.2.3
+	//
+	// +optional
+	HttpLogFormat string `json:"httpLogFormat,omitempty"`
+}
+```
+
+The `LoggingDestination` type is a union type:
+
+```go
+// LoggingDestination describes a destination for log messages.
+// +union
+type LoggingDestination struct {
+	// type is the type of destination for logs.  It must be one of the
+	// following:
+	//
+	// * Container
+	//
+	// Access logs are written to a sidecar container for each ingress
+	// controller pod replica.  The ingress operator configures the sidecar.
+	// The expectation is that the administrator configures a custom logging
+	// solution that reads logs from this sidecar.
+	//
+	// * Syslog
+	//
+	// Access logs are sent to a syslog endpoint.  The administrator must
+	// specify an endpoint that can receive syslog messages.  The
+	// expectation is that the administrator has configured a custom syslog
+	// instance.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type LoggingDestinationType `json:"type"`
+
+	// syslog holds parameters for a syslog endpoint.  Present only if
+	// type is Syslog.
+	//
+	// +optional
+	Syslog *SyslogLoggingDestinationParameters `json:"syslog,omitempty"`
+
+	// container holds parameters for the Container logging destination.
+	// Present only if type is Container.
+	//
+	// +optional
+	Container *ContainerLoggingDestinationParameters `json:"container,omitempty"`
+}
+```
+
+The union discriminator can have either the value `Container` or the value
+`Syslog`:
+
+```go
+// LoggingDestinationType is a type of destination to which to send log
+// messages.
+//
+// +kubebuilder:validation:Enum=Container;Syslog
+type LoggingDestinationType string
+
+const (
+	// Container sends log messages to a sidecar container.
+	ContainerLoggingDestinationType LoggingDestinationType = "Container"
+
+	// Syslog sends log messages to a syslog endpoint.
+	SyslogLoggingDestinationType LoggingDestinationType = "Syslog"
+)
+```
+
+For `Syslog`, an endpoint must be specified, and a facility may be specified (if
+unspecified the default is "local1"); `Container` has no parameters:
+
+```go
+// SyslogLoggingDestinationParameters describes parameters for the Syslog
+// logging destination type.
+type SyslogLoggingDestinationParameters struct {
+	// endpoint identifies the syslog host:port endpoint that receives log
+	// messages.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Endpoint string `json:"endpoint"`
+
+	// facility specifies the syslog facility of log messages.
+	//
+	// If this field is empty, the facility is "local1".
+	//
+	// +optional
+	Facility string `json:"facility,omitempty"`
```

Comment:
Is this opaque? Are there any more strict validations we could apply, even in terms of length?